surefire-providers/common-junit48/src/main/java/org/apache/maven/surefire/common/junit48/MatchMethodDescriptions.java

```diff
@@ -0,0 +1,74 @@
+package org.apache.maven.surefire.common.junit48;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.junit.runner.Description;
+import org.junit.runner.manipulation.Filter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Only run test methods in the given failure set
+ *
+ * @author mpkorstanje
+ */
+final class MatchMethodDescriptions
+    extends Filter
+{
+
+    private final List<Filter> filters = new ArrayList<Filter>();
+
+    MatchMethodDescriptions( Iterable<Description> descriptions )
+    {
+        for ( Description description : descriptions )
+        {
+            filters.add( matchMethodDescription ( description ) );
+        }
+    }
+
+    @Override
+    public boolean shouldRun( Description description )
+    {
+        for ( Filter filter : filters )
```

Comment:
This loop is different from the deleted class but I need to have spare time to have deep looking.

Reply:
`MatchMethodDescriptions` is created with a list of failed test descriptions. FailingMethodFilter was created with a list of failed class-method tuples.

When `shouldRun` is called the `FailingMethodFilter` would recurse down the description tree to find a leave (a test case).  For each leave it uses the class of the description to find the set of failed methods for that class. Then it tests if the set of failed methods contain the method of the description. If so it returns true and the test should be rerun.

When `shouldRun` is called the `MatchMethodDescriptions` will check for each failed-test-description if it can be found in the the tree of descriptions. This is done by delegating the recursive descent search to `Filter#matchMethodDescription`. `matchMethodDescription` recurses until it finds a leave, when found it compares descriptions.

This should explain where the recursion has been removed.